tezos/encoding-derive/src/gen.rs

```diff
@@ -0,0 +1,237 @@
+            GeneratorKind::FactoryWithEncoding(ty, span) => {
+                let method = format_ident!("{}", ty);
+                let encoding = super::enc::generate_encoding(encoding);
+                quote_spanned!(span=> f.#method(&(#field_name), #encoding))
+            }
+            GeneratorKind::FactoryWithEncoding2(ty, encoding2, span) => {
+                let method = format_ident!("{}", ty);
+                let encoding = super::enc::generate_encoding(encoding);
+                let encoding2 = super::enc::generate_encoding(encoding2);
+                quote_spanned!(span=> f.#method(&(#field_name), #encoding, #encoding2))
+            }
+            GeneratorKind::GeneratedType(ty) => {
+                quote!(#ty::generator(&(#field_name), f))
+            }
+        }
+    }
+}
+
+fn decompose_encoding<'a>(
+    encoding: &'a Encoding<'a>,
+    prefix: TokenStream,
+) -> (
+    GeneratorKind<'a>,
+    Option<(&'a Encoding<'a>, &'static str, TokenStream)>,
+) {
+    match encoding {
+        Encoding::Primitive(primitive, span) => (
+            GeneratorKind::FactoryPrimitive(
+                get_primitive_number_mapping(*primitive).unwrap(),
+                *span,
+            ),
+            None,
+        ),
+        Encoding::Bytes(span) => (GeneratorKind::FactoryPrimitive("u8", *span), None),
+        Encoding::String(_, span) => (GeneratorKind::FactoryWithEncoding("string", *span), None),
+        Encoding::Path(path) => (GeneratorKind::GeneratedType(path), None),
+        Encoding::List(_, encoding, span) => (
+            GeneratorKind::FactoryWithEncoding2("size", encoding, *span),
+            Some((encoding, "vec_of_items", quote!(#prefix + "[]"))),
+        ),
+        Encoding::Sized(_, encoding, span) => (
+            GeneratorKind::FactoryWithEncoding2("size", encoding, *span),
+            Some((encoding, "vec_of_items", quote!(#prefix + "[]"))),
+        ),
+        Encoding::OptionField(encoding, span) => (
+            GeneratorKind::FactoryPrimitive("bool", *span),
+            Some((encoding, "option", quote!(#prefix + "?"))),
+        ),
+        Encoding::Bounded(_, encoding, _) => decompose_encoding(encoding, prefix),
+        Encoding::Dynamic(_, encoding, _) => decompose_encoding(encoding, prefix),
+        _ => unreachable!("not supported by decompose_encoding: {:?}", encoding),
+    }
+}
+
+fn generator_for_encoding(encoding: &Encoding, prefix: TokenStream) -> TokenStream {
+    let (method_kind, decomposed) = decompose_encoding(encoding, prefix.clone());
+    let generator = method_kind.generate(encoding, prefix);
+    if let Some((inner_encoding, compose_func, prefix)) = decomposed {
+        let compose_func = format_ident!("{}", compose_func);
+        let inner_generator = generator_for_encoding(inner_encoding, prefix);
+        quote!(tezos_encoding::generator::#compose_func(#inner_generator, #generator))
+    } else {
+        generator
+    }
+}
+
+lazy_static::lazy_static! {
```

Comment:
Is using https://doc.rust-lang.org/std/lazy/struct.Lazy.html instead of `lazy_static` an option here? I have been thinking that we should start switching to that.